linchpin/shell/__init__.py

```diff
@@ -323,6 +328,38 @@ def fetch(ctx, fetch_type, remote, root):
         sys.exit(1)
 
 
+@runcli.command()
+@click.option('-o', '--outputfile', metavar='outputfile',
+              default=None, type=click.Path(),
+              help='Output file path to be written \
+                    if not mentioned will be redirected to stdout')
+@click.option('-f', '--format', metavar='format', default="cfg",
+              help='Inventory output format')
+@click.option('-t', '--tx-id', metavar='tx_id', type=int, default=None,
+              help='Transaction ID to be used to generate inventory')
+@click.option('-ot', '--output-type', metavar='output_type',
+              default='inventory', required=False,
+              help='Type of output to be generated,\
+                    currently supports inventory only')
+@click.option('-p', '--pos', metavar='pos', default="-1",
+              help='If multiple inventories mention which inventory\
+                    to be display. By default latest inventory is\
+                    displayed')
+@pass_context
+def generate(ctx, outputfile, format, tx_id, output_type, pos):
+    ret_bool = lpcli._write_to_inventory(inv_path=outputfile, inv_format=format)
+    if pos == "all":
+        for inventory in ret_bool:
+            click.echo(inventory)
+    else:
+        try:
+            click.echo(ret_bool[int(pos)])
+        except IndexError as e:
+            click.echo("Invalid index of inventory generated")
+            click.echo(e.message)
+    return ret_bool
+
+
```

Comment:
@herlo : Currently linchpin's rundb does not store inventory. Further , I have designed linchpin generate such that user can have a option to generate provider specific inventory by mentioning --provider-type in future . Currently linchpin generate has multiple options of specifying --inventory-format , --inventory-type and --outputfile Adding these three options to linchpin journal in addition to the ones which already have felt overwhelming it with many options.
^^^ Any thoughts?

Reply:
@samvarankashyap what does 'generate' mean here? It feels like it 'generates' something, but what? Adding an new subcommand means documenting it and detailing what is involved with the 'generate'.  If we were to do this, I'd argue that at a minimum, the subcommand should be named appropriately. Something like 'inventory' makes more sense in that case. Additionally, If the inventory were stored, other parameters would not be possible, so I see why you chose to make it a separate subcommand. 

On the other hand, 'journal' provides a 'a transactional record' for each task, so we're just taking the data from the record and displaying it in different ways. In this way, the journal is useful to a user.  As you stated in your comment above, the inventory doesn't need to be stored in the rundb, only the topology and layout. Using the topology, layout, and possibly other dynamic data, one could also create an inventory using the 'journal' subcommand. However, one could also obtain the linchpin.latest, or create a shell file that could be sourced to create environment variables. The possibilities are fairly unlimited. 

The vision I had for this was something akin to a single subcommand that could use either a tx_id or run_id/target combination, pass in a few options, and get an output to either STDOUT or a file with exactly the data a user would want/need. Something like the following:

```
$ linchpin journal --tx-id 82 --template-data @./domain-data.yml \
   --output-type=json-inventory
```
The above command would print json to STDOUT. Alternatively, one might want a linchpin.latest with shell environment values inside.

```
$ linchpin journal --target=aws-ec2-cluster --output-file /tmp/tmpfile.2.env \
   --output-type env-latest

```
The list is not limited here.

```
$ linchpin journal -tx-id 84 --target=aws-ec2 --output-type json-distilled \
   -o /tmp/tmpfile.3.distilled --fields='aws_ec2=instances.id,instances.tags:name'
```

What you can see here is that while at runtime, the `linchpin up` command should be able to output these sorts of items, and `linchpin journal` should be able to replicate them. Naming the subcommand `generate` or `journal` doesn't really matter per se. Adding a new subcommand seems like overkill to me, since journal already does the heavy lifting to look up the record data.

The only real adjustment needed would be to take what the generate subcommand does now and adapt it to the above. 

If you have a better name for 'journal', I'd be open to leaving what is there and adjusting to that name. Documentation aside, I believe this functionality is all related to looking up a record, and returning some sort of output a user (or application) can use.

Reply:
@samvarankashyap I did have one other thought for you. I was looking at the difference in outputs in rundb. Between 'resources' and 'topology_outputs', the only real difference I can see is that topology_outputs adds the structure around it for each different resource definition. In my mind, that's a lot of possible duplicate data. 

Would it be advantageous to change the 'resources' component to have the same structure as topology outputs? It might mean changing the lookups a little in the API, but then we wouldn't need to duplicate data in the rundb.

A snippet to help point the issue might be useful here:

```
"resources": [
    {
        "changed": true,
        "failed": false,
        "instance_ids": [
            "i-0881adb4d84c4e470",
            "i-0de04151d162add7b",
            "i-0e320bf9ab05092bd"
        ],
        "instances": [
            {
                "ami_launch_index": "2",
                "architecture": "x86_64",
                "block_device_mapping": {
                    "/dev/sda1": {
                        "delete_on_termination": true,
                        "status": "attached",
                        "volume_id": "vol-032b73e03f8458215"
                    },
                    "/dev/sdb": {
                        "delete_on_termination": true,
                        "status": "attached",
                        "volume_id": "vol-0ca7d8e99e5c33976"
```
vs
```
"topology_outputs": {
   "aws_cfn_res": [],
   "aws_ec2_key_res": [],
   "aws_ec2_res": [
       {
           "changed": true,
           "failed": false,
           "instance_ids": [
               "i-0881adb4d84c4e470",
               "i-0de04151d162add7b",
               "i-0e320bf9ab05092bd"
           ],
           "instances": [
               {
                   "ami_launch_index": "2",
                   "architecture": "x86_64",
                   "block_device_mapping": {
                       "/dev/sda1": {
                           "delete_on_termination": true,
                           "status": "attached",
                           "volume_id": "vol-032b73e03f8458215"
```
I'm not sure why it matters here as it seems maybe you could use the 'resources' section for this, and I know we discussed this previously. However, duplication of data seems a waste for space. Why not consolidate this into one structure? The reason the resources is a list is so we can include multiple different resource topologies in one target, but I'm not opposed to having the data in the structure you use for topology_outputs (just resources reformatted).

Reply:
@herlo : @p3ck 
After a code walk through I realized following: 
resources --> format is being used by distiller and linchpin.latest file generation ( which I see referenced in couple of independent function calls). 
topology_outputs --> format is being is used by inventory generation. 
consolidating both  of the formats involves re-writing logic for ( distiller, linchpin.latest ) or inventory generation. In my opinion it would be definitely be advantageous to switch to topology_outputs structure since it has more data w.r.to providers. 

In order to keep the code changes minimal . I had to sought for duplication of the data. 
If we would like to go ahead and following topology_outputs format ( which I personally think is easier to implement) . I might take a couple of days of effort. To make sure its not breaking any of distiller/linchpin.latest file. Would we mind if the PR is delayed with the requested changes ?

Reply:
The only qualm I have is that we keep the name 'resources', as topology_outputs is a long way of saying the same thing. Beyond that, I don't know which is better. Merging this PR, then making changes to another one is also fine as this one works. @p3ck do you have any suggestions on this? It does feel like the 'generate' -> 'journal' functionality would need to be in this PR. But since we just started a new sprint and aren't releasing anytime soon, I'm could go either way with it as long as the functionality is adjusted.

Reply:
I think a little delay is acceptable.  It sounds like the right approach to me.